himbaechel/uarch/gowin/pack.cc

```diff
@@ -0,0 +1,1373 @@
+        }
+        make_iob_nets(*out_iob);
+    }
+
+    void pack_single_output_iol(CellInfo &ci, std::vector<IdString> &nets_to_remove)
+    {
+        IdString out_port = id_Q;
+
+        CellInfo *out_iob = net_only_drives(ctx, ci.ports.at(out_port).net, is_iob, id_I, true);
+        NPNR_ASSERT(out_iob != nullptr && out_iob->bel != BelId());
+        BelId iob_bel = out_iob->bel;
+
+        BelId l_bel = get_iologic_bel(out_iob);
+        if (l_bel == BelId()) {
+            log_error("Can't place IOLOGIC %s at %s\n", ctx->nameOf(&ci), ctx->nameOfBel(iob_bel));
+        }
+        // mark IOB as used by IOLOGIC
+        out_iob->setParam(id_IOLOGIC_IOB, 1);
+        check_iologic_placement(ci, ctx->getBelLocation(iob_bel), out_iob->attrs.count(id_DIFF_TYPE));
+
+        if (!ctx->checkBelAvail(l_bel)) {
+            log_error("Can't place %s at %s because it's already taken by %s\n", ctx->nameOf(&ci),
+                      ctx->nameOfBel(l_bel), ctx->nameOf(ctx->getBoundBelCell(l_bel)));
+        }
+        ctx->bindBel(l_bel, &ci, PlaceStrength::STRENGTH_LOCKED);
+        std::string out_mode;
+        switch (ci.type.hash()) {
+        case ID_OVIDEO:
+            out_mode = "VIDEORX";
+            break;
+        case ID_OSER10:
+            out_mode = "ODDRX5";
+            break;
+        }
+        ci.setParam(ctx->id("OUTMODE"), out_mode);
+
+        // disconnect Q output: it is wired internally
+        nets_to_remove.push_back(ci.getPort(out_port)->name);
+        out_iob->disconnectPort(id_I);
+        ci.disconnectPort(out_port);
+        set_daaj_nets(ci, iob_bel);
+
+        Loc io_loc = ctx->getBelLocation(iob_bel);
+        if (io_loc.y == ctx->getGridDimY() - 1) {
+            config_bottom_row(*out_iob, io_loc, Bottom_io_POD::DDR);
+        }
+        make_iob_nets(*out_iob);
+    }
+
+    IdString create_aux_iologic_name(IdString main_name, int idx = 0)
+    {
+        std::string sfx("");
+        if (idx) {
+            sfx = std::to_string(idx);
+        }
+        return ctx->id(main_name.str(ctx) + std::string("_aux") + sfx);
+    }
+
+    BelId get_aux_iologic_bel(const CellInfo &ci)
+    {
+        return ctx->getBelByLocation(gwu.get_pair_iologic_bel(ctx->getBelLocation(ci.bel)));
+    }
+
+    bool is_diff_io(BelId bel) { return ctx->getBoundBelCell(bel)->attrs.count(id_DIFF_TYPE) != 0; }
+
+    CellInfo *create_aux_iologic_cell(CellInfo &ci, IdString mode, bool io16 = false, int idx = 0)
+    {
+        if (ci.type.in(id_ODDR, id_ODDRC, id_OSER4, id_IDDR, id_IDDRC, id_IDES4)) {
+            return nullptr;
+        }
+        IdString aux_name = create_aux_iologic_name(ci.name, idx);
+        BelId bel = get_aux_iologic_bel(ci);
+        BelId io_bel = gwu.get_io_bel_from_iologic(bel);
+        if (!ctx->checkBelAvail(io_bel)) {
+            if (!is_diff_io(io_bel)) {
+                log_error("Can't place %s at %s because of %s\n", ctx->nameOf(&ci), ctx->nameOfBel(bel),
```

Comment:
can you add a short reason to this error message, something like `because of a conflict with %s` rather than just `because of %s`?